vault/plugin_catalog.go

```diff
@@ -0,0 +1,169 @@
+package vault
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"path/filepath"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/vault/helper/builtinplugins"
+	"github.com/hashicorp/vault/helper/jsonutil"
+	"github.com/hashicorp/vault/helper/pluginutil"
+	"github.com/hashicorp/vault/logical"
+)
+
+var (
+	pluginCatalogPrefix = "plugin-catalog/"
+)
+
+// PluginCatalog keeps a record of plugins known to vault. External plugins need
+// to be registered to the catalog before they can be used in backends. Builtin
+// plugins are automatically detected and included in the catalog.
+type PluginCatalog struct {
+	catalogView *BarrierView
+	directory   string
+
+	lock sync.RWMutex
+}
+
+func (c *Core) setupPluginCatalog() error {
+	c.pluginCatalog = &PluginCatalog{
+		catalogView: c.systemBarrierView.SubView(pluginCatalogPrefix),
+		directory:   c.pluginDirectory,
+	}
+
+	return nil
+}
+
+// Get retrieves a plugin with the specified name from the catalog. It first
+// looks for external plugins with this name and then looks for builtin plugins.
+// It returns a PluginRunner or an error if no plugin was found.
+func (c *PluginCatalog) Get(name string) (*pluginutil.PluginRunner, error) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	// Look for external plugins in the barrier
+	out, err := c.catalogView.Get(name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve plugin \"%s\": %v", name, err)
+	}
+	if out != nil {
+		entry := new(pluginutil.PluginRunner)
+		if err := jsonutil.DecodeJSON(out.Value, entry); err != nil {
+			return nil, fmt.Errorf("failed to decode plugin entry: %v", err)
+		}
+
+		return entry, nil
+	}
+
+	// Look for builtin plugins
+	if factory, ok := builtinplugins.BuiltinPlugins.Get(name); ok {
+		return &pluginutil.PluginRunner{
+			Name:           name,
+			Builtin:        true,
+			BuiltinFactory: factory,
+		}, nil
+	}
+
+	return nil, fmt.Errorf("no plugin found with name: %s", name)
+}
+
+// Set registers a new external plugin with the catalog, or updates an existing
+// external plugin. It takes the name, command and SHA256 of the plugin.
+func (c *PluginCatalog) Set(name, command string, sha256 []byte) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	parts := strings.Split(command, " ")
+	command = parts[0]
+	args := parts[1:]
+
+	command = filepath.Join(c.directory, command)
+
+	// Best effort check to make sure the command isn't breaking out of the
+	// configured plugin directory.
+	sym, err := filepath.EvalSymlinks(command)
+	if err != nil {
+		return fmt.Errorf("error while validating the command path: %v", err)
+	}
+	symAbs, err := filepath.Abs(filepath.Dir(sym))
+	if err != nil {
+		return fmt.Errorf("error while validating the command path: %v", err)
+	}
+
+	if symAbs != c.directory {
```

Comment:
We are already calling `filepath.Abs` on the plugin directory in core: 
https://github.com/hashicorp/vault/pull/2200/files#diff-ca11aecd476e9dc90c4f0643b402dddfR470